### PR TITLE
Introduce "module.xrun"

### PR DESCRIPTION
**Pull request by isbm**
### What does this PR do?

Adds a feature: `module.xrun` for states.

### Motivation

Current way of calling Salt modules with `module.run` has pretty cumbersome interface along with the fact it has also reserved words. This leads to a strange states, something like:

```yaml
show_friends:
  module.run:
    - name: facebook.list_friends
    - m_name: thatch
```

Or even worse:

```yaml
list_states:
  module.run:
    - name: united.states
    - m_states: Florida, Arizona, California
    - m_names: Tallahassee, Phoenix, Sacramento
    - m_fun: Bowling
    - kwargs: {
      'drink_beer': True,
      'visit': [
        'Florida State Capitol',
        'Arizona Science Center',
        'California Railroad State Museum'
      ],
      'call': thatch
    }
```

This led to the reasoning make it easier and simpler to use, throw away `m_` prefix (this just sucks, sorry). Also write better unit tests. :smile: So the above can be done something like this:

```yaml
show_friends:
  module.xrun:
    facebook.list_friends:
      - name: thatch
```

Look, Ma! No `m_` prefix! How about another example:

```yaml
list_states:
  module.xrun:
    united.states:
      - states: Florida, Arizona, California
      - names: Tallahassee, Phoenix, Sacramento
      - fun: Bowling
      - drink_beer: True
      - visit:
        - Florida State Capitol
        - Arizona Science Center
        - California Railroad State Museum
      - call: thatch
```

**One more thing** you can do (optionally, of course, you don't _have_ to):

```yaml
this_calls_it_at_once:
  module.xrun:
    some.function:
      - foo
      - bar
    another.function:
      - wow
      - that's handy
```

:sunglasses: 

### Tests written?

Yes

### CODE REVIEW

**Review by thatch45 — APPROVED** [review]
This looks great, I never liked the module.run interface. I think this is significantly cleaner

**Comment by isbm** [comment]
@thatch45 thanks! :) Before @cachedout pointed me to the Lint, I am already fixing it!!

**Review by gtmanfred — CHANGES_REQUESTED** [review]
:+1:

**Review by gtmanfred — APPROVED** [review]

**Comment by isbm** [comment]
@cachedout I do not understand Lint error of W1699 against `zip(...` and on top of the source. It works just fine in Python3 and Python2. What exactly is wrong and how to fix it?

**Review by rallytime — COMMENTED** [review]
Very nice! This will be great to have.

I only have one small request: Can you add this new option to the [Nitrogen Release Notes](https://github.com/saltstack/salt/blob/develop/doc/topics/releases/nitrogen.rst)? I think people will find this useful and will want to know about it!

**Review by whiteinge — APPROVED** [review]
Might be worth adding a deprecation notice to `module.run` for this release or the next.

Given how heavily used it is we might not want to actually put it on a deprecation path but rather just add a notice to that docstring -- the 'ol `file.sed` treatment. 😉

**Comment by isbm** [comment]
@rallytime I added [release notes](https://github.com/isbm/salt/blob/9e30aa331f8a98583c3f964aa2f28fa8113c2b6c/doc/topics/releases/nitrogen.rst#state-module-changes) and hopefully it is correct.

@whiteinge I added decorator-based deprecation (and also fixed it is now working for state modules as well). Now it will simply be using new `module.run`, but you can always restore an old behaviour with the `use_deprecated` mechanism. I mentioned all that in examples within the release notes.

**Comment by isbm** [comment]
@cachedout by now all the issues were addressed. If you think those two lints are just misbehaviour of custom scripts (at my side it doesn't hurt for Python 3 actually), then we are clean to merge.

**Comment by whiteinge** [comment]
@isbm sorry my comment above wasn't very clear.

I strongly think that we should **not** deprecate the old `module.run`. It's been around for far too long and is used in the wild far too much. The "file.sed treatment" is adding a note to the docstring of the old function that says, "Hey, you should use <this> other function instead," but keeping the old one in the codebase more or less indefinitely.

Thoughts from the Core team? Agree/disagree?

**Comment by isbm** [comment]
@whiteinge But this is what exactly is happening. Add:
```yaml
use_deprecated:
  - module.run
```
And you will have the old one. All is needed to be done is to package it properly. E.g. we will ship something like `/etc/salt/minion.d/module_run_deprecated.conf` with the content above. Then you will see those messages in log and you can just "fix" the syntax eventually. Because having _both_ that does the same is kind of pollution (in my eyes). Having that renamed is to lose "run" name, which nothing can be better.

So let's just slowly migrate the world. Two years should be enough to write `- foo: bar` instead of `{foo: 'bar'}`. :smile:

**Comment by thatch45** [comment]
Both good points, I think that slowly moving the world would be a good idea, we could easily make sure that the comment returned from module.run specifies that it is going to be deprecated...

**Comment by whiteinge** [comment]
I can't speak directly to pros/cons of relying on package upgrades to put that config file in place. Personally, I'd prefer an opt-in rather than an opt-out for such a core module.

That said, I'll defer to the Core team since they will be on the front lines of any user upgrade problems.

**Comment by thatch45** [comment]
I would also prefer for opt-in, we do not have the luxury that SUSE has of dictating how the platform is going to be used unfortunately :(

**Comment by isbm** [comment]
@whiteinge @thatch45 no problems for opt-in. I will change that in the entire mechanism, so all the deprecation will work always on opt-in with "nag messages" to the log. Good point! But now is 1:00AM in Germany, so let me fix that tomorrow CEST. :-)

**Comment by isbm** [comment]
@thatch45 [specially for your preference](https://github.com/saltstack/salt/pull/39891/commits/ea5f86e84422f4b078c440d46acc0737e242e58c)! Have a nice weekend, guys.

File: salt/states/module.py
Comment (ScoreUnder) [inline salt/states/module.py]: Could potentially overwrite the `ret['comment']` & `ret['result']` from a previous loop iteration where it fails, depending on the content of the SLS file and the order of the dict.

Comment (isbm) [inline salt/states/module.py]: Oh, knapp, you're right. And I forgot to update docstring. Give me a sec...

**Comment by isbm** [comment]
@ScoreUnder done!

File: doc/topics/releases/nitrogen.rst
Comment (terminalmage) [inline doc/topics/releases/nitrogen.rst]: Should be "has dropped".

File: doc/topics/releases/nitrogen.rst
Comment (terminalmage) [inline doc/topics/releases/nitrogen.rst]: This can be expressed better in RST using:

```
.. note::
    It is necessary to explicitly turn on the new behavior (see below)
```

File: doc/topics/releases/nitrogen.rst
Comment (terminalmage) [inline doc/topics/releases/nitrogen.rst]: @thatch45 Did we ever make a firm decision on whether to use US or international spelling for words ending in ``-or``/``-our``?

File: doc/topics/releases/nitrogen.rst
Comment (terminalmage) [inline doc/topics/releases/nitrogen.rst]: This should be

```
Please keep in mind that the old syntax will no longer be supported in the ``Oxygen`` release of Salt.
```

File: doc/topics/releases/nitrogen.rst
Comment (terminalmage) [inline doc/topics/releases/nitrogen.rst]: The note here could also be put in a ``.. note::`` RST block.

I don't understand this note though, to be honest. Are you suggesting that all minions should have their config files edited? Have you considered using [config.get](https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.config.html#salt.modules.config.get) to get these configuration options? It checks grains and pillar as well as the minion config file. This would allow for this option to be turned on simply by editing the pillar data (or using ``grains.setval`` to set the value in the minion's /etc/salt/grains). Both of these options seem less intrusive IMO than editing all minions' config files.

File: doc/topics/releases/nitrogen.rst
Comment (terminalmage) [inline doc/topics/releases/nitrogen.rst]: This would read better as:

```
To enable the new behavior, add the following to the minion config file:
```

**Review by terminalmage — CHANGES_REQUESTED** [review]
I think the config options can be better retrieved using ``config.get``. There are also a number of things I noticed in the RST that we can clean up.

File: doc/topics/releases/nitrogen.rst
Comment (isbm) [inline doc/topics/releases/nitrogen.rst]: Not american here, sry.

File: doc/topics/releases/nitrogen.rst
Comment (isbm) [inline doc/topics/releases/nitrogen.rst]: @terminalmage good point on grains. I will extend that, but with a different PR, since I would like to have a different tests for it.

File: doc/topics/releases/nitrogen.rst
Comment (terminalmage) [inline doc/topics/releases/nitrogen.rst]: I honestly don't care which we use, but I think that we should choose one for consistency across the docs.

Comment (cachedout) [inline doc/topics/releases/nitrogen.rst]: They are mixed in the docs. I agree we should pick a style but since we haven't, I see no reason to block on this.

**Comment by cachedout** [comment]
Final call for comments on this one. Otherwise, everything is looking good for a merge.

**Comment by rallytime** [comment]
@cachedout What do you want to do about the PY3 lint errors for the changes in `module.py`?

https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/9449/violations/file/salt/states/module.py/

**Comment by isbm** [comment]
@rallytime I already asked this: please advice what it does mean. Because the code is **not** incompatible with Python 3.

**Comment by rallytime** [comment]
@isbm Right, all good points. 

For record, and so everyone doesn't have to click around and find it, the console output for the PY3 violation displays the following:
```
16:14:34 salt/states/module.py:183: [W1699(incompatible-py3-code), ] Incompatible Python 3 code found. Proposed fix:
16:14:34  from salt.ext.six.moves import range
16:14:34  from salt.exceptions import SaltInvocationError
16:14:34  from salt.utils.decorators import with_deprecated
16:14:34 +from salt.ext.six.moves import zip
16:14:34  
16:14:34  
16:14:34  def wait(name, **kwargs):
16:14:34 salt/states/module.py:311: [W1699(incompatible-py3-code), ] Incompatible Python 3 code found. Proposed fix:
16:14:34      :return:
16:14:34      '''
16:14:34      argspec = salt.utils.args.get_function_argspec(__salt__[name])
16:14:34 -    func_kw = dict(zip(argspec.args[-len(argspec.defaults or []):], argspec.defaults or []))
16:14:34 +    func_kw = dict(list(zip(argspec.args[-len(argspec.defaults or []):], argspec.defaults or [])))
16:14:34      func_args = []
16:14:34      for funcset in kwargs.get('func_args') or {}:
16:14:34          if isinstance(funcset, dict):
```
But, as you stated, this should be working in both versions. If this is something wrong with our linter, then we need to address that.

I am thinking @s0undt3ch or @cachedout should provide some direction on this.

**Comment by isbm** [comment]
@rallytime Hmm. I don't think this is needed fix: the only difference that `zip` in Python3 returns iterator instead of list. This fix is trying to replicate the behaviour in a sense that it first turns iterator to the list. I personally wouldn't do that.

**Comment by cachedout** [comment]
@s0undt3ch Can you comment?

**Comment by isbm** [comment]
I sense this branch will become incompatible and @cachedout will ask me to rebase... :laughing:

**Comment by cachedout** [comment]
@isbm Heh. Just trying to get feedback all resolved. We'll get this in as soon as we can. :]

File: doc/topics/releases/nitrogen.rst
Comment (terminalmage) [inline doc/topics/releases/nitrogen.rst]: In that case, yes I definitely agree.

**Comment by terminalmage** [comment]
The flagged code is definitely valid, ``zip()`` will happily accept and iterator on both Python 2 and 3. I think we can skip this test with a ``# pylint: disable=incompatible-py3-code`` at the end of that line. Remember, two spaces after the end of the line or else the code will fall afoul of a different pylint check. :smile:

**Comment by isbm** [comment]
@terminalmage thanks. Done.

File: salt/states/module.py
Comment (cachedout) [inline salt/states/module.py]: Pylint isn't this smart I'm afraid. This needs to go up one line.

Comment (isbm) [inline salt/states/module.py]: @cachedout amended.

**Comment by isbm** [comment]
@cachedout another final call? :laughing:

**Comment by terminalmage** [comment]
There are a bunch of lint failures still, but they seem to be due to @s0undt3ch's changes in the test suite to keep us from using distutils.version.LooseVersion due to PY3 issues.

A rebase should clear those up and let us know if any legitimate lint failures remain.

**Comment by cachedout** [comment]
@isbm Is thie branch totally in-sync with Salt's upstream? These failures don't look related but it's not clear to me where they're coming from.

**Comment by isbm** [comment]
@cachedout OK, today I re-based it one more time against `origin/develop` for you. Lint errors at least gone and unit tests works. What else I should do?

**Comment by cachedout** [comment]
@isbm That's all I think we need. If the tests pass, I'll merge it.

**Comment by isbm** [comment]
@cachedout it is all @rallytime 's fault. :laughing: (no, not really: thanks for adding unit tests for `with_deprecated` which supposed to be _mine_ technical debt!)

**Pull request merged**
